HPCC-35269 Add unittests to XRef

jackdelv opened the pull request:
- Refactor XRef to make it usable with the CPPUNIT framework. I wasn't sure of the best way to refactor or where to move the code. After some back and forth, I settled on removing all of the XRef classes from saxref.cpp and creating a subfolder under sasha with the XRef functionality.
- Add unittests in xrefunittests.cpp and export them as a library. I also export the XRef classes as a library to be used elsewhere.
- Add option to image.sh for allowing unittests to be built. I am not sure if this was the best way to add that functionality, but it works.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith reviewed (changes requested):
@jackdelv - I think sensible to refactor in general, so it is easier to add cppunit tests.
But I think it would be better to leave saxref.cpp should be left largely untouched, and the cppunit tests to live next to the code, and have a new xreflib.cmake or similar.
May make sense to compile the other sasha services into the xref lib too (if we add unittests for code in other services it would make sense).
It would be good to mimimize the churn. Hopefully this way the new unit tests themselves would be the main diff.

jakesmith commented:
>and have a new xreflib.cmake or similar.

or, since dali/sasha/CMakeLists is already building both saserver and the 'sasha' CLI, you could add the xreflib rules there too.

jackdelv committed:
HPCC-35269 Add unittests to saxref.cpp

jackdelv commented:
@jakesmith I reverted my migration of the xref code and left saxref.cpp mostly intact except for the unittests and added to the dali/sasha/CMakeLists.txt to export saxref and its dependencies as a shared library. I also added a command-line option to image.sh for turning on the unittests in containerized.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -3203,3 +3203,445 @@ void runExpiryCLI()
+        XRefAllocator allocator(1); // 1MB limit
+
+        // Try to allocate more than the limit
+        try {
```
trivial/formatting: brace on newline

jakesmith reviewed (commented):
@jackdelv - I think looks a lot better.

> have a new xreflib.cmake or similar.
> May make sense to compile the other sasha services into the xref lib too (if we add unittests for code in other services it would make sense).

You'd have to make a few other small changes I think, but the new libary could be used/linked by saserver I think, to avoid duplicating the code into the lib and exe.
Did you look what minimal changes that would involve?

jakesmith reviewed (changes requested):
@jackdelv - I think this should wait for the separate PR that refactors sasha service code into a lib. See https://github.com/hpcc-systems/HPCC-Platform/pull/20961

jakesmith reviewed (commented):
@jackdelv - please see comments.

jakesmith reviewed (commented):
@jackdelv - looks good, 1 minor question.

Also the JIRA description should be filled in, and the title of the JIRA, PR and commit message should be amended , as this is now adding unittets not refactoring XREF.

jackdelv commented on dali/sasha/CMakeLists.txt:
```diff
@@ -107,6 +107,11 @@ target_link_libraries ( saserver
 
+if (USE_CPPUNIT)
+    target_link_libraries ( saserver cppunit )
+    include_directories( ${HPCC_SOURCE_DIR}/testing/unittests)
```
I was including the unittests directory because I was using unittests.hpp to be able to use the CPPUNIT assert functions. I removed it and directly included <cppunit/extensions/HelperMacros.h>.

jakesmith reviewed (changes requested):
@jackdelv - please address build issues.

and re. https://github.com/hpcc-systems/HPCC-Platform/pull/20590#pullrequestreview-3831491678 ... can you update the JIRA description please.

jakesmith reviewed (commented):
@jackdelv - lokos good. Plesae squash.

jakesmith reviewed (approved):
@jackdelv - looks good.
